src/container/srv_container.c

```diff
@@ -1647,196 +1649,248 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl, struct cont *cont,
+			if (eph_ldr->ea_server_ephs[i].re_stable_eph < stable_eph)
+				eph_ldr->ea_server_ephs[i].re_stable_eph = stable_eph;
 			break;
 		}
 	}
 
-	if (i == ec_agg->ea_servers_num) {
+	if (i == eph_ldr->ea_servers_num) {
 		if (!retried) {
-			D_DEBUG(DB_MD, "rank %u eph "DF_X64" retry for"
-				DF_CONT"\n", rank, eph,
+			D_DEBUG(DB_MD, "rank %u ec_agg_eph "DF_X64", stable_eph "DF_X64
+				" retry for"DF_CONT"\n", rank, ec_agg_eph, stable_eph,
 				DP_CONT(pool_uuid, cont_uuid));
 			retried = true;
-			ec_agg->ea_deleted = 1;
+			eph_ldr->ea_deleted = 1;
 			goto retry;
 		} else {
-			D_WARN("rank %u eph "DF_X64" does not exist for "
-			       DF_CONT"\n", rank, eph,
+			D_WARN("rank %u ec_agg_eph "DF_X64", stable_eph "DF_X64
+			       " does not exist for "DF_CONT"\n", rank, ec_agg_eph, stable_eph,
 			       DP_CONT(pool_uuid, cont_uuid));
 		}
 	} else {
-		D_DEBUG(DB_MD, DF_CONT" update eph rank %u eph "DF_X64"\n",
-			DP_CONT(pool_uuid, cont_uuid), rank, eph);
+		D_DEBUG(DB_MD, DF_CONT" update eph rank %u ec_agg_eph "DF_X64
+			", stable_eph "DF_X64".\n", DP_CONT(pool_uuid, cont_uuid),
+			rank, ec_agg_eph, stable_eph);
 	}
 
 out_put:
 	cont_svc_put_leader(svc);
 	return 0;
 }
 
-struct refresh_vos_agg_eph_arg {
-	uuid_t	pool_uuid;
-	uuid_t  cont_uuid;
-	daos_epoch_t min_eph;
+#define EPH_ARG_TGT_INLINE	(64)
+struct refresh_track_eph_arg {
+	uuid_t		 pool_uuid;
+	uuid_t		 cont_uuid;
+	daos_epoch_t	 min_ec_agg_eph;
+	daos_epoch_t	 min_stable_eph;
+	uint8_t		*tgt_status;
+	uint8_t		 tgt_status_inline[EPH_ARG_TGT_INLINE];
 };
 
-int
-cont_refresh_vos_agg_eph_one(void *data)
+static int
+cont_refresh_track_eph_one(void *data)
 {
-	struct refresh_vos_agg_eph_arg *arg = data;
-	struct ds_cont_child	*cont_child;
-	int			rc;
+	struct refresh_track_eph_arg	*arg = data;
+	struct ds_cont_child		*cont_child;
+	unsigned int			 idx = dss_get_module_info()->dmi_tgt_id;
+	int				 rc;
 
 	rc = ds_cont_child_lookup(arg->pool_uuid, arg->cont_uuid, &cont_child);
 	if (rc)
 		return rc;
 
-	D_DEBUG(DB_MD, DF_CONT": %s agg boundary eph "DF_X64"->"DF_X64"\n",
+	D_DEBUG(DB_MD, DF_CONT": %s ec agg boundary eph "DF_X64"->"DF_X64", "
+		": %s stable eph "DF_X64"->"DF_X64"\n",
 		DP_CONT(arg->pool_uuid, arg->cont_uuid),
-		cont_child->sc_ec_agg_eph_boundary < arg->min_eph ? "update" : "ignore",
-		cont_child->sc_ec_agg_eph_boundary, arg->min_eph);
-
-	if (cont_child->sc_ec_agg_eph_boundary < arg->min_eph)
-		cont_child->sc_ec_agg_eph_boundary = arg->min_eph;
+		cont_child->sc_ec_agg_eph_boundary < arg->min_ec_agg_eph ? "update" : "ignore",
+		cont_child->sc_ec_agg_eph_boundary, arg->min_ec_agg_eph,
+		cont_child->sc_stable_epoch < arg->min_stable_eph ? "update" : "ignore",
+		cont_child->sc_stable_epoch, arg->min_stable_eph);
+
+	if (cont_child->sc_ec_agg_eph_boundary < arg->min_ec_agg_eph)
+		cont_child->sc_ec_agg_eph_boundary = arg->min_ec_agg_eph;
+
+	/* Only should update local stable epoch if the target is in UPIN status */
```

Comment:
For the target with `PO_COMP_ST_UP` status, should we consider its local stable epoch?

Reply:
here it is setting local's persistent global stable epoch, only UPIN node can set it. 
For reint it should rebuild from the local stored stable epoch before exclude it (still be UPIN).

The UP tgt reports its local epoch by query vos_cont_get_local_stable_epoch() in cont_aggregate_interval().

Reply:
yes, there is a time window between excluding and every tgt server get the new pool map.
But as long as this tgt is still UPIN in current poolmap, it will report its local stable epoch and on PS leader will ensure the global stable epoch will not exceed every UPIN tgt.
So here even in the time window that this engine with stable pool map still safe to update its local stored global stable epoch, right?

Reply:
hmm, on leader side, if a rank is DOWN, I think it'll be ignored when leader finding out the lowest stable epoch, right?

Reply:
right, that case is - leader with latest pool map that tgt A is DOWN, will ignore its reports then global stable epoch can exceed A's local stable epoch.
But A itself with stable poolmap still think itself is UPIN.
But when leader try to refresh the stable epoch by cont_iv_track_eph_refresh(), the IV msg propogate will find group version mismatch and return -DER_GRPVER, and will retry the IV OP until grp ver matched on all nodes.
So far looks fine?

Reply:
I see, I think we are good as long as the IV request with different pool map version can be rejected, thanks for the explanation.